### PR TITLE
(Upstream) Adds docx & fb3 support, some glyph caching tweaks

### PR DESCRIPTION
**Pull request by poire-z**
Includes 2 commits from https://github.com/buggins/coolreader/pull/111 by @virxkane (other commits are not relevant because we diverged in these part of the code and don't have the bug they solve).
First commit cleans up glyph caching by removing some dedicated code, which makes thing cleaner.
Second commit adds back hash table stuff that was used in that removed dedicated code - but its use is disabled as it does not provide any speedup.

Includes FB3 & DOCX support from https://github.com/buggins/coolreader/pull/107 by @pkb , with a few tweaks because of a few things we do differently:
- don't include fb3.css, as it just adds a few stuff to fb2.css - so add them to fb2.css.
The changes are small, but the diff is huge, but I had to reformat it from dos to unix format because git complained about some mismatch line endings I couldn't solve some other way... It's only this:
```diff
+ol title, ul title {
+       page-break-before: auto;
+}

-b,strong,i,em,dfn,var,q,u,del,s,strike,small,big,sub,sup,acronym,tt,sa mp,kbd,code {
+b,strong,i,em,dfn,var,q,u,underline,del,s,strike,spacing,small,big,sub,sup,acronym,tt,sa mp,kbd,code {
    display: inline;
 }

-u { text-decoration: underline; }
+u, underline { text-decoration: underline; }
 del, s, strike, strikethrough { text-decoration: line-through; }
+spacing { letter-spacing: 5px; }
```

- I first adapted and included docx.css (which is more FB2/crengine tweaks oriented), but ended up removing it so we don't have to manage another CSS file from frontend - and I wanted to have it work with epub.css with classic HTML elements so our style tweaks can be used
- ~~for that, I had to tweak how titles and footnotes are made into the DOM, so the last 2 commits - hope that won't make it too hard to follow upstream improvments. (I had a hard time with the headings. I wanted to just have `<h2>`, or even `<section><h2>` instead of `<section><title class=h2><p>` - but I was missing some closing tags so all my 200-pages test document was in imbricated H1 and made 50000 pages ... So ended up just having `<section><title><h2>`...)~~
This has been integrated by upstream, and provided as an alternative option, that we just enable.

To be bumped after next release.
Will need into thirdparty/kpvcrlib/CMakeLists.txt:
```diff
+    ${CRE_DIR}/src/lvopc.cpp
+    ${CRE_DIR}/src/docxfmt.cpp
+    ${CRE_DIR}/src/fb3fmt.cpp
```
and into frontend/document/credocument.lua:
```diff
     self.default_css = "./data/epub.css"
-    if file_type == "fb2" then
+    if file_type == "fb2" or file_type == "fb3" then
         self.default_css = "./data/fb2.css"
     end
[...]
+    registry:addProvider("docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document", self, 90)
+    registry:addProvider("fb3", "application/fb3", self, 90)
```

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]
Looks good, also the DOCX stuff seems to be quite self-contained so I read through it fairly superficially (but given that you adapted it you probably didn't).

**Comment by pkb** [comment]
> I had a hard time with the headings. 

@poire-z why didn't you just tell me. I wrote the code recently so it is easier for me to change it. I've extracted title handling into a separate class. To have HTML like DOM you just need to comment/uncomment appropriate title handler in ImportDocXDocument() thus all the difference is localized, we shouldn't have issues with future changes if there will be any :). Please let me know if the DOM structure generated is what you expect?

**Comment by poire-z** [comment]
@pkb Oh, thanks :)
Well, I didn't want to put the burden of adapting it for us on you :)

The DOM structure is fine, and the footnotes and emphasis>em changes are good.
You might want to wrap this into some `#ifdef` like you did with `DOCX_CRENGINE_IN_PAGE_FOOTNOTES`
```c
    //Two options when dealing with titles: (FB2|HTML)
    docx_fb2TitleHandler titleHandler(&writer, &importContext); //<section><title class=hx">..</title></section>
    //docx_titleHandler titleHandler(&writer, &importContext);  //<hx>..</hx>
    docx_documentHandler documentHandler(&docReader, &writer, &importContext, &titleHandler);
```

But there is some issue: we get no TOC built when using `docx_titleHandler`.

I initially went with plain `<Hx>` tags and no `<section><title>` wrapper (so, no auto TOC by crengine section>title mechanism), so I had added some code to rebuild the TOC from document headings - [we already have that in koreader's crengine](https://github.com/koreader/crengine/blob/71dd8583cf9b1bd09e6efd906834ebc2521b59f3/crengine/src/lvtinydom.cpp#L14705-L14756), it just needed a thin wrapper so it can be called at the end of ImportDocXDocument:
```c
void ldomDocument::buildTocFromHeading()
{
    m_toc.clear(); // the footnotes body>section would still make it into the TOC
    getRootNode()->recurseElements(makeTocFromCrHintsOrHeadings);
}
```

But as I had some issues with non closing Hx tags, and went back to using your section>title>hX, the TOC was built by that crengine mechanism, so I removed my added TOC stuff as it wasn't needed anymore.

I really didn't mind your SECTION>TITLE wrapping my Hx, as for us, SECTION and TITLE are just transparent block elements with no style (so no added margin/padding).

So, I don't know how you want it:
- current limited docx_titleHandler as it would not build a TOC (and let that to the users/me that want it)
- current limited docx_titleHandler, but add a streamlined version of our `makeTocFromCrHintsOrHeadings` so you can use it to build the TOC
- have docx_titleHandler still output the SECTION>TITLE wrapper
- some other idea to build the TOC without SECTION>TITLE :)

(You might want to remove any mention of "upstream" in my comments that you picked - only us should use that word to mention you :)

**Comment by pkb** [comment]
> You might want to wrap this into some #ifdef

Done.  Not sure if define name make sense, though :)

I think you may use current limited docx_titleHandler and add makeTocFromCrHintsOrHeadings() to build TOC if it works for you. I'll leave it as it is now in my branch. Maybe @buggins will say his opinion about this when time will come.

**Comment by poire-z** [comment]
OK, thanks, I'll do that :)

**Comment by pkb** [comment]
I did one more commit. Now it is possible to control docx_fb2TitleHandler with DOCX_USE_CLASS_FOR_HEADING define. If it is set to true class attribute will be generated in title element and if false <hx> tag in SECTION>TITLE wrapper.  Now all options covered. I hope, I did not messed up during all this refactoring.

**Comment by poire-z** [comment]
@pkb: it does matter in any thing, but for correctness, you could add that:
```diff
     ldomDocumentWriter writer(doc);
     docXMLreader docReader(&writer);

+#ifdef DOCX_FB2_DOM_STRUCTURE
     writer.OnStart(NULL);
     writer.OnTagOpen(NULL, L"?xml");
     writer.OnAttribute(NULL, L"version", L"1.0");
     writer.OnAttribute(NULL, L"encoding", L"utf-8");
     writer.OnEncoding(L"utf-8", NULL);
     writer.OnTagBody();
     writer.OnTagClose(NULL, L"?xml");
     writer.OnTagOpenNoAttr(NULL, L"FictionBook");
     // DESCRIPTION
     writer.OnTagOpenNoAttr(NULL, L"description");
     writer.OnTagOpenNoAttr(NULL, L"title-info");
     writer.OnTagOpenNoAttr(NULL, L"book-title");
     writer.OnTagClose(NULL, L"book-title");
     writer.OnTagClose(NULL, L"title-info");
     writer.OnTagClose(NULL, L"description");
+#else
+    writer.OnStart(NULL);
+    writer.OnTagOpen(NULL, L"?xml");
+    writer.OnAttribute(NULL, L"version", L"1.0");
+    writer.OnAttribute(NULL, L"encoding", L"utf-8");
+    writer.OnEncoding(L"utf-8", NULL);
+    writer.OnTagBody();
+    writer.OnTagClose(NULL, L"?xml");
+    writer.OnTagOpenNoAttr(NULL, L"html");
+#endif

 #ifdef DOCX_FB2_DOM_STRUCTURE
     //Two options when dealing with titles: (FB2|HTML)
     docx_fb2TitleHandler titleHandler(&writer, &importContext); //<section><title class=hx">..</title></section>
 #else
     docx_titleHandler titleHandler(&writer, &importContext);  //<hx>..</hx>
 #endif
```

**Comment by poire-z** [comment]
> Now it is possible to control docx_fb2TitleHandler with DOCX_USE_CLASS_FOR_HEADING define

So, you're saying I may prefer to use docx_fb2TitleHandler , with that define set to false, to get both my Hx and the TOC ? :)

**Comment by poire-z** [comment]
So, with your latest code, and all my stuff not included, with just:
```diff
 // comment this out to disable in-page footnotes
-#define DOCX_CRENGINE_IN_PAGE_FOOTNOTES 1
+// #define DOCX_CRENGINE_IN_PAGE_FOOTNOTES 1
 // build FB2 DOM, comment out to build HTML DOM
 #define DOCX_FB2_DOM_STRUCTURE 1
 //If true <title class="hx"><p>...</p></title> else <title><hx>..</hx></title>
-#define DOCX_USE_CLASS_FOR_HEADING true
+#define DOCX_USE_CLASS_FOR_HEADING false
```
it works fine with us (just as with my stuff in this PR). We get the same DOM, with a nice TOC:

<kbd>image</kbd>

With:
```diff
 // comment this out to disable in-page footnotes
-#define DOCX_CRENGINE_IN_PAGE_FOOTNOTES 1
+//#define DOCX_CRENGINE_IN_PAGE_FOOTNOTES 1
 // build FB2 DOM, comment out to build HTML DOM
-#define DOCX_FB2_DOM_STRUCTURE 1
+//#define DOCX_FB2_DOM_STRUCTURE 1
 //If true <title class="hx"><p>...</p></title> else <title><hx>..</hx></title>
 #define DOCX_USE_CLASS_FOR_HEADING true

```
We get a cleaner DOM, but no TOC:
<kbd>image</kbd>

So, we'd need to add some small bit of code to make a TOC after the DOM is built.

I'd probably go with the first solution, as I don't mind as much about the DOM for DocX as I do for EPUB :) and I don't plan any work on DocX, so better to stay close to upstream default options.
@Frenzie : any opinion?

**Comment by Frenzie** [comment]
The first looks fine to me. Skipping out on some boilerplate in the generated DOM - eh, whatever? :-)

**Comment by poire-z** [comment]
Well, the 2nd would make it easier to anyone wanting to style a bit more the basic styles of docx with style tweaks.
With the 1st one, simple/classic CSS like `H1 + P { text-indent: 0; }` would not work, while it would on the 2nd. One would need a docx-dedicated  and more fuzzy `TITLE + P {text-indent: 0;}` tweak.
(But as the docx hierarchy and elements set is quite limited, one could also make a full stylesheet dedicated to docx, docx.css, to its liking.)

**Comment by Frenzie** [comment]
I think no TOC would be more immediately felt by your average user but I doubt it matters much. I'd say just pick the one with simpler code/DOM if that's what you prefer.

**Comment by poire-z** [comment]
OK, finally went with the 2nd solution: clean HTML DOM, + TOC built from headings after the DOM is made.

**Comment by Frenzie** [comment]
Alright, looks good to me.

**Comment by Frenzie** [comment]
> Glyph caching: optional use of hash table instead of a linked list
> By @virxkane, from https://github.com/buggins/coolreader/pull/111
> 5th commit adapted for our non-splitted lvfntman.cpp.
> 
> Disabled by default, as it does not provide any noticable
> performance boost.

I was looking around for some potential quick wins and I saw a 25-30% speed increase in drawing (turning) pages. Of course that's talking about mere milliseconds for something that's quite fast regardless, so not nearly as big as it sounds, but I figured it was worth noting. Also it may be different for CJK. Maybe a question of interest to @spfenwick?
